fs/io_zenfs.cc

```diff
@@ -632,6 +633,171 @@ size_t ZonedRandomAccessFile::GetUniqueId(char* id, size_t max_size) const {
+// traffic activity via the PCIe channel to the ZNS SSD because
+// of the read/write to zones which needs  to be issued. We need
+// some better ideas later to bring in efficiency, something like
+// "simple copy" or ideas in those lines.
+IOStatus ZenFSGCWorker::MoveValidDataToNewDestZone() {
+  std::vector<Zone*>::iterator zone_it;
+  std::vector<ZoneExtent*>::iterator ext_it;
+  std::vector<ZoneExtent*>::iterator e_it;
+  IOStatus s;
+  uint64_t r_pos;
+  uint32_t size;
+  uint32_t long_ext_size;
+  uint64_t new_start;
+  const char* ptr;
+  int dont_read = 0;
+
+  // Sort the Extent list in decreasing order.
+  std::sort(extent_list.begin(), extent_list.end(),
+            [](ZoneExtent* ext1, ZoneExtent* ext2) {
+              return ext1->length_ > ext2->length_;
+            });
+
+  // Get the size of the largest extent.
+  long_ext_size = extent_list[0]->length_;
+
+  // Allocate a buffer with size of the largest extent.
+  // We have to issue pread from the source zones so we
+  // need a buffer.
+  ptr = new char[long_ext_size];
+
+  zone_it = dst_zone_list.begin();
+  for (ext_it = extent_list.begin(); ext_it != extent_list.end();) {
+    ZoneExtent* ext;
+    Zone* zone_dst;
+
+    ext = *ext_it;
+    zone_dst = *zone_it;
+
+    // Set the position and length in the source zone to
+    // read the data.
+    r_pos = ext->start_;
+    size = ext->length_;
+    Slice buf(ptr, size);
+
+    if (!dont_read) s = ReadExtent(&buf, r_pos, ext->zone_);
+
+    // Free the allocated memory if error.
+    if (!s.ok()) {
+      if (ptr) delete[] ptr;
+      return s;
+    }
+
+    // Store the new starting position for the extent
+    // which will be later made persistent.
+    new_start = zone_dst->wp_;
+
+    // Write the valid data where were read from the
+    // source zone to the destination zone.
+    s = zone_dst->Append((char*)ptr, size);
+    if (s.ok()) {
+      // Data was written to the new zone, so the extent
+      // will have a new starting position. No need to
+      // change the length of the extent as it will be the
+      // same.
+      ext->start_ = new_start;
+
+      // The extent was moved to a new zone so change the
+      // resident zone parameter of the extent.
+      ext->zone_ = zone_dst;
+
+      // Current extent was written so now fetch the next extent.
+      ext_it++;
+      memset((char*)ptr, 0, long_ext_size);
+      continue;
+    }
+
+    if (s == IOStatus::NoSpace()) {
+      // Data was already read before, no need to read it again
+      dont_read = 1;
+
+      // The current zone cannot fit this extent because of lack
+      // of space, so get the next zone from the dst_zone_list.
+      zone_it++;
+    }
+
+    // There was an error so cannot proceed and simply
+    // we return the status.
+    if (s == IOStatus::IOError()) {
+      // If memory was allocated, we free it before returning.
+      if (ptr) delete[] ptr;
+
+      return s;
+    }
+  }
+
+  if (ptr) delete[] ptr;
```

Comment:
If I remembered correctly, `delete[] ptr` won't set ptr to nullptr, and there would be double free issue if there are two `delete[]` on the same code path.

Reply:
It seems that all code paths are covered and all of them only have one `delete[]`. We don't care whether `ptr` is null or not (it will always be non-null), so let's just remove `if (ptr)` condition.